Source/Core/AudioCommon/XAudio2Stream.cpp

```diff
@@ -219,6 +219,7 @@ void XAudio2::SetRunning(bool running)
     else
       m_voice_context->Stop();
   }
+  return true;
```

Comment:
Just something I noticed when going over it again, but shouldn't `false` be returned in the case `m_voice_context` is null? As in:

```cpp
bool XAudio2::SetRunning(bool running)
{
   if (!m_voice_context)
    return false;

   if (running)
     m_voice_context->Play();
   else
     m_voice_context->Stop();

  return true;
}
```

Seems odd to return true in the case it's not possible to actually do anything. Ditto for the XAudio2_7 SoundStream as well.

Reply:
Hmm I thought about that, but it doesn't seem possible to have an initialized XAudio sound stream without a voice context! Is it better to handle these defensive cases, or just assume that the object is already in a legitimate state?

This sort of is an issue with the two-stage init process in general; maybe it would be better to make the constructors private and have each subclass provide a `Create() -> std::optional<Subclass>` static, or something like that.

Reply:
@lioncash ping

Reply:
> it doesn't seem possible to have an initialized XAudio sound stream without a voice context

Right, but from an API perspective:

```cpp
auto ss = std::make_unique<Xaudio2>();
ss->SetRunning(true);
```

can be done and get the result of success.

Making the constructors private leaves you with ugly stuff like:

```cpp
std::unique_ptr<Type>(new Type(arguments, here));
```

since it makes `std::make_unique` unusable. Also `std::optional` can't be used in a polymorphic context. It's designed as a value type. You'll slice your object passing a `std::optional<Derived>` to a `std::optional<Base>`.

As in:

```cpp
class A
{
public:
  virtual ~A() = default;
  virtual void Print()
  {
    std::cout << "Base" << '\n';
  }
};

class B : public A
{
public:
  void Print() override
  {
    std::cout << "Derived" << '\n';
  }
};

static std::optional<A> Make(bool b)
{
  if (b)
    return std::make_optional<A>();

  return std::make_optional<B>();
}

int main()
{
  auto one = Make(false);
  auto two = Make(true);

  one->Print();
  two->Print();

  return 0;
}
```

will print `Base` twice. So it kind of kills flexibility in this case (you'd need to put the full `std::unique_ptr` into the optional).

There's no real cut-and-dry/one-size-fits-all answer for what should be done, since it depends on what is preferable. The ISO C++ Core Guidelines do have a discussion similar to something like this though in their [discussion](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#S-discussion) portion of the guidelines, though. Specifically, the one titled "Discussion: Use a factory function if you need “virtual behavior” during initialization", which might help out more.

Reply:
@lioncash Okay, so you're saying that it's better to be defensive here instead of saying "if you don't check the result of Start/Init, then other methods are UB"? I'm fine with that.

> `std::optional` isn't advisable in a direct polymorphic context.

Right, I meant you'd return a non-sliced `std::optional<Subclass>` and push off the conversion to a polymorphic SoundStream until later. Or just return a `std::unique_ptr<SoundStream>` (conflating constructor success w/ allocation, but less verbose).

I did see the guidelines discussion on factory functions, though from my reading it offers a number of tradeoffs rather than a specific solution. It admits that a big issue with factory functions is the added verbosity and non-standard construction syntax. What I was really asking in the previous comment was: what do you want to see here, in order to approve the PR?

Reply:
@lioncash ping

Reply:
I was only really suggesting the simple boolean check in my initial comment for the meantime.

Ideally we'd have a better init pattern. IMO it's kind of silly to have both a constructor and also require an init function call after the fact in an API. Establishing invariants in a constructor and providing a way to query error status seems nicer/less clunky from a usability perspective; that, or attempting to lazily initialize internally in a way that doesn't explicitly require doing so, but that might just be me. However I *dooon't* quite think the purpose of this PR is to explore and restructure initialization in its entirety ;)

Reply:
I agree that's it's kind of silly 😄